src/main/java/org/springframework/data/mapping/FactoryMethod.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2011-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.mapping;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * Value object to encapsulate the factory method to be used when mapping persistent data to objects.
+ *
+ * @author Mark Paluch
+ * @since 3.0
+ */
+public final class FactoryMethod<T, P extends PersistentProperty<P>> extends EntityCreatorSupport<T, P> {
```

Comment:
Wonder if we could get as far as avoiding having to expose these types (just as `PreferredConstructor`) publically, now that `EntityCreator` is available.

Reply:
Depends on the design really. A `FactoryMethod`/`PreferredConstructor` makes only sense when there's reflective access. With our separation where we externalize `EntityInstantiator`s we always leak implementation details. `EntityCreator` reduces the concept of entity creation to the actual description of input parameters without actually defining the approach (reflection, constructor/factory method, a provided lambda) how entities are being created. Exposing `Executable` would already assume that we use reflection information in some sense to create entities.

Reply:
I get that, but most places in which the reflection API is used are not for the actual (reflective) invocation of those methods, but *the inspection* and decisions triggered by the outcome of that. Even in an AOT scenario, we would need that, wouldn't we? It feels pretty brittle to see all those instance of checks in different places, as they stop working as soon as you wrap the original instance in some decorator. The entire detection process leading to the creation of `FactoryMethod`/`PreferredConstructor` is reflection-based. If we avoid reflection happening in front of those instances, it's much easier to potentially replace the entire detection process and instances. For example, by an instance of `EntityCreator` that will just obtain that metadata from a different place without using *any* runtime reflection at all.